Dashboard improv: add button to filter for the user's org only

shivam-Purohit opened the pull request:
Button to switch between org only reports or all the reports 
closes #1936

# Description


## Type of change


- [x] New feature (non-breaking change which adds functionality).


# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [ ] A new plugin (analyzer, connector, visualizer, playbook, pivot or ingestor) was added or changed, in which case:
    - [ ] I strictly followed the documentation ["How to create a Plugin"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-add-a-new-plugin)
    - [ ] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](./Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [ ] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [ ] I have dumped the configuration from Django Admin using the `dumpplugin` command and added it in the project as a data migration. [Doc](["How to create a Plugin"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-add-a-new-plugin))
    - [ ] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [ ] If you created a new analyzer and it is free (does not require API keys), please add it in the `FREE_TO_USE_ANALYZERS` playbook in `playbook_config.json`.
    - [ ] Check if it could make sense to add that analyzer/connector to other [freely available playbooks](https://intelowl.readthedocs.io/en/develop/Usage.html#list-of-pre-built-playbooks).
    - [ ] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
- [ ] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](./Contribute.md)).
- [x] If the GUI has been modified:
    - [ ] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.

### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.

shivam-Purohit commented:
@mlodic its almost done now. But the problem is the response is cached so if I change the users the previous response is shown not the new one.

shivam-Purohit commented:
this is how the UI looks like. I am not good with names. We can get the switch which has label inside the switch only. That looks way cooler though we will need to add one more library for that.
![ubuntuMachine  Running  - Oracle VM VirtualBox 22-12-2023 13_05_45](https://github.com/intelowlproject/IntelOwl/assets/91889807/9f995423-018e-4ddd-b98c-ede6b24adcae)

mlodic commented:
hello @shivam-Purohit, thank you for your PR! :)

> @mlodic its almost done now. But the problem is the response is cached so if I change the users the previous response is shown not the new one.

when you say "if I change the users", you mean if you logout and then login with another user?

> We can get the switch which has label inside the switch only. That looks way cooler though we will need to add one more library for that.

Do you mean something like that? 
![image](https://github.com/intelowlproject/IntelOwl/assets/30625432/b9fc2f4d-a4c8-4ec1-9f6b-522c65f4bfdd)

This how we switch the visualization in the "Job Result" page. It could make sense to leverage the same component also here. Maybe, I'd also put that component at the left of the time picker, not at the right.

Thoughts?

mlodic commented on api_app/views.py:
```diff
@@ -478,22 +479,238 @@ def aggregate_md5(self, request):
+    ) -> Response:
         delta, basis = self.__parse_range(self.request)
+        filter_kwargs = {"received_request_time__gte": delta}
+        if user:
+            filter_kwargs["user"] = user
+        if users:
+            filter_kwargs["user__in"] = users
```
I think you can optimize the code by choosing a single parameter "users" here and change the filter based on the number of items in the list

mlodic commented on api_app/views.py:
```diff
@@ -478,22 +479,238 @@ def aggregate_md5(self, request):
+            # Handle the case where the user has no membership
+            logger.info("User has no membership.")
+
+        if organization:
+            return self.__aggregation_response_dynamic(
+                "md5", False, users=users_of_organization
+            )
+        else:
+            return self.__aggregation_response_dynamic("md5", False, user=current_user)
```
we should call the aggregation method a single time, we need to optimize this code

mlodic commented on api_app/views.py:
```diff
@@ -478,22 +479,238 @@ def aggregate_md5(self, request):
+        current_user = request.user
+        organization = None
+        users_of_organization = None
+        try:
+            if current_user.has_membership():
+                organization = current_user.membership.organization
+                if organization:
+                    users_of_organization = [
+                        membership.user for membership in organization.members.all()
+                    ]
+
+        except AttributeError:
+            # Handle the case where the user has no membership
+            logger.info("User has no membership.")
```
this code has been replicated lots of time. We need it just once

mlodic commented on frontend/src/components/dashboard/utils/charts.jsx:
```diff
@@ -20,19 +20,38 @@ import {
-export const JobStatusBarChart = React.memo(() => {
+export const JobStatusBarChart = React.memo((props) => {
   console.debug("JobStatusBarChart rendered!");
+  /* eslint-disable */
+  var sendURL = JOB_AGG_STATUS_URI;
+  const porp = props.myprop;
+  const getValue = porp.key;
+  console.debug("props is ", getValue);
+  /* eslint-enable */
+
+  if (getValue) {
+    /* eslint-disable */
+    sendURL = JOB_AGG_STATUS_URI_ORG;
+    console.debug("this is sending to my urls", sendURL);
+    /* eslint-enable */
+  }
```
more generally, I don't think it is necessary to create new API endpoints for this case. We should use the URL parameters to customize the results. This would avoid a lot of additional code that would be really difficult to maintain.

Example:
aggregate/status -> like it is right now
aggregate/status?org=True -> with the toggle enabled

shivam-Purohit replied:
Yeah I was thinking the same. I just tried some sub-optimised code to check the core functionality. I will improve the code quality.

mlodic reviewed (changes requested):
there is some duplicated code that must be addressed

shivam-Purohit commented:
> when you say "if I change the users", you mean if you logout and then login with another user?

yes same. To check / test the feature I created two more users and added one to my org. Though when i change the users  i.e logout and login again I get the results from the previous user. But if we restart the container the new results do re-render. Either it caches the result or there is some bug in my code?

shivam-Purohit commented:
> This how we switch the visualization in the "Job Result" page. It could make sense to leverage the same component also here. Maybe you can call the 2 options: "Org / Global" and make it "toggable" only if the user is part of an org. Otherwise it could be hidden or disabled.
> Maybe, I'd also put that component at the left of the time picker, not at the right.

yes this look better. I will try that. But to disable the option if user is not part of any org we will need to send request to the backend during the rendering of dashboard.
if so then we wont need to check if user is part of any org in backend.

mlodic commented:
> > This how we switch the visualization in the "Job Result" page. It could make sense to leverage the same component also here. Maybe you can call the 2 options: "Org / Global" and make it "toggable" only if the user is part of an org. Otherwise it could be hidden or disabled.
> > Maybe, I'd also put that component at the left of the time picker, not at the right.
> 
> yes this look better. I will try that. But to disable the option if user is not part of any org we will need to send request to the backend during the rendering of dashboard. if so then we wont need to check if user is part of any org in backend.

yep I think there is already something similar in place for this case in the "plugin" section:
![image](https://github.com/intelowlproject/IntelOwl/assets/30625432/6860ad20-fca8-4609-b877-fe08c680a7ef)
The button related to org's config is shown only if the user belongs to an org

shivam-Purohit commented:
![ubuntuMachine  Running  - Oracle VM VirtualBox 28-12-2023 12_47_19](https://github.com/intelowlproject/IntelOwl/assets/91889807/8778ad4e-2d6e-42b0-b7f2-bf992648bed4)
![ubuntuMachine  Running  - Oracle VM VirtualBox 28-12-2023 11_49_46](https://github.com/intelowlproject/IntelOwl/assets/91889807/0778fedb-fdda-45d7-a3dc-83f932e00c38)

shivam-Purohit commented:
hey @mlodic hope the Christmas went well! 
i re-adjusted some code
changed the component to only render if the organisation exist for the user using `useOrganisationStore `components.
renamed new variables
and shifted some repeated code to new function

shivam-Purohit commented:
> ![ubuntuMachine Running - Oracle VM VirtualBox 28-12-2023 12_47_19](https://private-user-images.githubusercontent.com/91889807/293157026-8778ad4e-2d6e-42b0-b7f2-bf992648bed4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3NDgxNjgsIm5iZiI6MTcwMzc0Nzg2OCwicGF0aCI6Ii85MTg4OTgwNy8yOTMxNTcwMjYtODc3OGFkNGUtMmQ2ZS00MmIwLWI3ZjItYmY5OTI2NDhiZWQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI4VDA3MTc0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlZTRlZDFjMDkxMzg2ZGZhZWQ3ZmQ4NGMzZmQ2NGFhNDM3MjI4YTYzZWIxMDM3YzY0ZmJmNmM3ZDkwMDBiMjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.w2RHdnuNcBW4aDRJLVXWLM4BxYOaW2B212wD0iIYwjw) ![ubuntuMachine Running - Oracle VM VirtualBox 28-12-2023 11_49_46](https://private-user-images.githubusercontent.com/91889807/293156844-0778fedb-fdda-45d7-a3dc-83f932e00c38.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3NDgxNjgsIm5iZiI6MTcwMzc0Nzg2OCwicGF0aCI6Ii85MTg4OTgwNy8yOTMxNTY4NDQtMDc3OGZlZGItZmRkYS00NWQ3LWEzZGMtODNmOTMyZTAwYzM4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI4VDA3MTc0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3YmYzOWQ3YTMxNmI1NTIyNmIyZDdmNTZhYTM0Zjk2NTNkOGY1ZjcyZGVhNTU4YTgwOWNmOWE0Y2I1MmNmYTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GiZ9zy4WUNlKlhTXCVG4v1-iOvAgllnXPGl-okXTVMI)

this is how it looks. I have to change some layout to fit in new structure. If I just move the `buttongroup` component above the `timepicker ` in like the original code it was breaking like the buttongroup was moving to the left so i separated both the title and this components (timepicker and buttongroup).

mlodic commented:
hey, I hope this message finds you well :)

Unluckily I don't see the images that you linked in the last message. Could you please re-share them?

mlodic commented:
hey, I hope this message finds you well :)

I saw the images, that feels good!

mlodic commented on api_app/views.py:
```diff
@@ -60,7 +60,7 @@
 )
 
-logger = logging.getLogger(__name__)
+logger = logging.getLogger("__name__")
```
why this change?

shivam-Purohit replied:
I changed it while I was trying to add a console logger. I was having some problem with log file to see the recent changes. I will revert it back.

mlodic commented on api_app/views.py:
```diff
@@ -416,11 +416,17 @@ def pivot(self, request, pk=None, pivot_config_pk=None):
     )
     @cache_action_response(timeout=60 * 5)
     def aggregate_status(self, request):
+        org_param = request.GET.get("org", "").lower() == "true"
+        users_of_organization = None
+        if org_param:
+            users_of_organization = self.get_org_members(request.user)
```
this could be a method returning `users_of_organization`, right?

shivam-Purohit replied:
yes returns all the members in the organisation

mlodic replied:
I meant that those 4 lines could be a method which returns `users_of_organization`.

I am telling you this because I saw the code repeated multiple times

mlodic replied:
@shivam-Purohit onceyou get some time, if you could just do this little refactor then we can merge this PR :) thank you

shivam-Purohit replied:
Sorry i missed the comment above earlier. I think the new refactor should be good.

mlodic reviewed (commented):
cool to me, do you want to make a little review @drosetti too?

drosetti commented on frontend/src/components/dashboard/Dashboard.jsx:
```diff
@@ -57,20 +61,91 @@ export default function Dashboard() {
   return (
     <Container fluid id="Dashboard">
-      <div className="g-0 d-flex align-items-baseline flex-column flex-lg-row mb-2">
+      {/* <div className="g-0 d-flex align-items-baseline flex-column flex-lg-row mb-2">
         <h3 className="fw-bold" id="Dashboard_title">
           Dashboard
         </h3>
+        <ButtonGroup className="ms-2 mb-3 d-flex align-items-center">
+          <Button
+            outline={!state}
+            color={state ? "primary" : "tertiary"}
+            onClick={() => handleChange()}
+          >
+            GLOBAL
+          </Button>
+          <Button
+            outline={state}
+            color={!state ? "primary" : "tertiary"}
+            onClick={() => handleChange()}
+          >
+              ORG
+          </Button>
+        </ButtonGroup>
         <ElasticTimePicker
           className="ms-auto"
           size="sm"
           defaultSelected={range}
           onChange={onTimeIntervalChange}
           id="Dashboard_timepicker"
         />
+
+      </div> */}
```
remove this comment

drosetti commented on frontend/src/components/dashboard/Dashboard.jsx:
```diff
@@ -57,20 +61,91 @@ export default function Dashboard() {
+            <ButtonGroup className="mb-3">
+              <Button
+                outline={orgstate}
+                color={!orgstate ? "primary" : "tertiary"}
```
I suggest to reverse the check, i think it's easier to understand:
`orgstate ? "tertiary" : "primary"`

drosetti commented on frontend/src/components/dashboard/Dashboard.jsx:
```diff
@@ -39,8 +39,12 @@ const charts2 = [
+  // const isSelectedUI = JobResultSections.VISUALIZER;
   const { guideState, setGuideState } = useGuideContext();
 
+  const [orgstate, setorgState] = useState(() => false);
```
It's a very silly suggestion but I'd to have orgState, it has the same style of guideState (it's strange the linter doesn't mark this)

drosetti commented on frontend/src/components/dashboard/utils/charts.jsx:
```diff
@@ -27,12 +27,23 @@ const colors = getRandomColorsArray(10, true);
+  /* eslint-enable */
+
+  /* eslint-disable */
+  sendURL = `${JOB_AGG_STATUS_URI}?org=${getValue}`;
```
Pls use a more specific name for this const, what do you think about ORG_JOB_AGG_STATUS_URI ?

shivam-Purohit replied:
sure!

drosetti reviewed (changes requested):
I think you did a good job, I requested little changes (most of them on the names of the const/var). Be careful with `eslint-disable`, I know sometimes it's annoying, but it's useful to have same code style with many developers, it should be disabled only in rare cases.

shivam-Purohit commented:
@drosetti I have made the changes requested. One more thing I noticed is in the` aggregate_observable_classification`.Its counts the new ipv6 addressing IPs in the generic and not in the IP classification. That also reflect in the chart named `Job: Frequent IPs, Hash & Domains` which also does not count ipv6 IPs.
I can create a new issue regarding this and if it is not on higher priority I can fix that.

drosetti commented:
@shivam-Purohit I think this problem in the dashboard is related to the fact that the regex doesn't allow IPv6, then they are classified as generic and not show in the dashboard charts. Matteo told me you previously changed the regex to accept IPv6, I think I deleted it during a refactor, my apologies. However, I resolved and restored your work on the regex: in the develop branch you can find the refex that accept both IPv4 and IPv6, with a rebase everything should work, let me know if you have some problems. 

I also check your code and approved your pr from a frontend prospective, good work :muscle:

shivam-Purohit commented:
![ubuntuMachine  Running  - Oracle VM VirtualBox 05-01-2024 18_09_39](https://github.com/intelowlproject/IntelOwl/assets/91889807/79044bb8-c4a5-4daa-9d62-6f4f9afcc570)

@drosetti Thanks! I think we should be done here. Do let me know if further changes are required.
